packages/ra-core/src/form/useFormIsDirty.ts

```diff
@@ -0,0 +1,49 @@
+import { useFormState } from 'react-hook-form';
+import isEmpty from 'lodash/isEmpty.js';
+
+// useFormState().isDirty might differ from useFormState().dirtyFields (https://github.com/react-hook-form/react-hook-form/issues/4740)
+export const useFormIsDirty = (): boolean => {
```

Comment:
It's nice you extracted this function in a hook.
But, as I said, I recommend adding unit tests covering it, to ensure we support all use cases we expect (like scalar value, array of scalars, array of objects, etc.).
The goal being to capitalize on all the possible values you found by testing.